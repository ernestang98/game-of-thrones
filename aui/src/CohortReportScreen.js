import React, { useEffect } from 'react';
import {
    Typography,
    Box,
    Stack,
    Fab,
    List,
    ListItem,
    ListItemText,
    ListItemIcon,
} from "@mui/material";
import {Link} from "react-router-dom";
import MaterialTable from '@material-table/core'
import axios from 'axios';

import AddBox from "@material-ui/icons/AddBox";
import ArrowDownward from "@material-ui/icons/ArrowDownward";
import { forwardRef } from 'react';
import Check from '@material-ui/icons/Check';
import ChevronLeft from '@material-ui/icons/ChevronLeft';
import ChevronRight from '@material-ui/icons/ChevronRight';
import Clear from '@material-ui/icons/Clear';
import DeleteOutline from '@material-ui/icons/DeleteOutline';
import Edit from '@material-ui/icons/Edit';
import FilterList from '@material-ui/icons/FilterList';
import FirstPage from '@material-ui/icons/FirstPage';
import LastPage from '@material-ui/icons/LastPage';
import Remove from '@material-ui/icons/Remove';
import SaveAlt from '@material-ui/icons/SaveAlt';
import Search from '@material-ui/icons/Search';
import ViewColumn from '@material-ui/icons/ViewColumn';

import LocalFireDepartmentIcon from '@mui/icons-material/LocalFireDepartment';
import OneKPlusIcon from '@mui/icons-material/OneKPlus';

import NavigationIcon from '@mui/icons-material/Navigation';

/**
 * render Cohort Report Screen
 * @returns {JSX.Element}
 * @constructor
 */
function IndivReportScreen() {

    const tableIcons = {
        Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
        Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
        Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
        Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
        DetailPanel: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
        Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
        Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
        Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
        FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
        LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
        NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
        PreviousPage: forwardRef((props, ref) => <ChevronLeft {...props} ref={ref} />),
        ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
        Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
        SortArrow: forwardRef((props, ref) => <ArrowDownward {...props} ref={ref} />),
        ThirdStateCheck: forwardRef((props, ref) => <Remove {...props} ref={ref} />),
        ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />)
    };

    const [cohortResults, setCohortResults] = React.useState([]);
    const [bestStudent, setBestStudent] = React.useState({
        /**
         * Student email
         */
        student_email: "",
        /**
         * Student's total points earned so far
         */
        points_earned: 0,
        /**
         * In-game's total earnable points
         */
        max_points_earnable: 0,
        status: 0,
        /**
         * Student name
         */
        name: "",
        /**
         * Student's total points earned so far
         */
        points: 0,
        /**
         * Student's in-game rank
         */
        rank: 0,
        /**
         * Student's in-game position
         */
        position: 0
    });

    /**
     * Retrieve entire cohort results and details
     */
    const getCohortResults = () => {
        const endpoint = process.env.REACT_APP_API_ENDPOINT === null ? "https://fastapi-ernestang98.cloud.okteto.net/" : process.env.REACT_APP_API_ENDPOINT
        axios.get(endpoint + "attempt/cohortresults/")
        .then( function(res) {
            let cohortArray = res.data;
            cohortArray.forEach( function(arrayItem) {
                arrayItem["percentage_correct"] = (( arrayItem["points_earned"] / arrayItem["max_points_earnable"] ) * 100).toFixed(2);
            })
            setCohortResults(cohortArray);
            getBestStudent();
        })
        .catch( function(err) {
            console.log(err);
        })
    }

    /**
     * Calculate and find best performing student in the cohort and set the state of best student
     */
    const getBestStudent = () => {
        var highest_points_earned = Math.max.apply(Math,cohortResults.map(function(o){return o.points_earned;}))
        console.log(highest_points_earned);
        var bestStudentObj = cohortResults.find(function(o){ return o.points_earned == highest_points_earned; });
        console.log(bestStudentObj);
        setBestStudent(bestStudentObj);
    }

    /**
     * Hook to retrieve cohort results before rendering of DOM
     */
    useEffect( ()=> {
        getCohortResults();
    })

    return (
            <Box sx={{ my: 5, justifyContent: "center", flexDirection: "row", alignItems: "center" }}>
                <Stack
                spacing={5}
                justifyContent="center"
                alignItems="center">
                     <Fab id="indiv-link" variant="extended" color="primary" component={Link} to={"/indiv-report"}>
                        <NavigationIcon sx={{ mr: 1 }} />
                        Individual Student Report
                    </Fab>
                    <Typography align="center" variant="h3">
                        Cohort Report
                    </Typography>
                    <Box
                     sx={{
                        width: 300,
                        // height: 300,
                    }}
                    >
                        <Box sx={{ width: '100%', maxWidth: 360, bgcolor: 'background.paper' }}>
                            <List>
                                <ListItem>
                                    <ListItemIcon>
                                        <LocalFireDepartmentIcon />
                                    </ListItemIcon>
                                    { bestStudent &&
                                        <ListItemText>
                                            Best Performing Student: {bestStudent.name}
                                        </ListItemText>
                                    }

                                </ListItem>
                                <ListItem>
                                    <ListItemIcon>
                                        <OneKPlusIcon />
                                    </ListItemIcon>
                                    { bestStudent &&
                                        <ListItemText>
                                            Total Points Earned: {bestStudent.points_earned}
                                        </ListItemText>
                                    }
                                </ListItem>
                            </List>
                        </Box>
                    </Box>


                    <Box
                    sx={{
                        width: "100%",
                        height: 300,
                    }}
                    >
                        <MaterialTable
                            columns={[
                                { title: 'Name', field: 'name' },
                                { title: 'Rank', field: 'rank', type: 'numeric' },
                                { title: 'Position', field: 'position', type: 'numeric' },
                                { title: 'Points Earned', field: 'points_earned', type: 'numeric' },
                                { title: 'Percentage Correct', field: 'percentage_correct', type: 'numeric' },

                            ]}
                            data={cohortResults}
                            title="Report"
                            options={{filtering:true}}
                            icons={tableIcons}
                        />
                    </Box>
                </Stack>
            </Box>

    );
}

export default IndivReportScreen