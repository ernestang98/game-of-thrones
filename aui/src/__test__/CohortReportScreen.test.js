import React from 'react';
import {mount, shallow} from 'enzyme';
import CohortReportScreen from "../CohortReportScreen";

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import "@testing-library/jest-dom/extend-expect"
import toJson from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });

describe("Cohort Report Screen", () => {
    it("render success", () => {
        const wrapper = shallow(<CohortReportScreen />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('link to cohort report screen', () => {
        const wrapper = shallow(<CohortReportScreen />);
        expect(wrapper
            .find('#indiv-link')
            .props()
            .to
            )
            .toBe('/indiv-report');
    })

    // it("should update selected student state on select student", () => {
    //     const selectStudent = jest.fn();
    //     const wrapper = shallow(<CohortReportScreen />);
    //     const handleClick = jest.spyOn(React, "useState");
    //     handleClick.mockImplementation(selectedStudent => [selectedStudent, setSelectedStudent]);
     
    //     wrapper.find("#demo-simple-select").simulate("click");
    //     expect(selectStudent).toBeTruthy();
    // });
});

