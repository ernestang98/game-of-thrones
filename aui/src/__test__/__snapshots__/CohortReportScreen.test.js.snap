// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cohort Report Screen render success 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "my": 5,
    }
  }
>
  <ForwardRef(Stack)
    alignItems="center"
    justifyContent="center"
    spacing={5}
  >
    <ForwardRef(Fab)
      color="primary"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      id="indiv-link"
      to="/indiv-report"
      variant="extended"
    >
      <Memo(ForwardRef(NavigationIcon))
        sx={
          Object {
            "mr": 1,
          }
        }
      />
      Individual Student Report
    </ForwardRef(Fab)>
    <ForwardRef(Typography)
      align="center"
      variant="h3"
    >
      Cohort Report
    </ForwardRef(Typography)>
    <ForwardRef(Box)
      sx={
        Object {
          "width": 300,
        }
      }
    >
      <ForwardRef(Box)
        sx={
          Object {
            "bgcolor": "background.paper",
            "maxWidth": 360,
            "width": "100%",
          }
        }
      >
        <ForwardRef(List)>
          <ForwardRef(ListItem)>
            <ForwardRef(ListItemIcon)>
              <Memo(ForwardRef(LocalFireDepartmentIcon)) />
            </ForwardRef(ListItemIcon)>
            <ForwardRef(ListItemText)>
              Best Performing Student: 
            </ForwardRef(ListItemText)>
          </ForwardRef(ListItem)>
          <ForwardRef(ListItem)>
            <ForwardRef(ListItemIcon)>
              <Memo(ForwardRef(OneKPlusIcon)) />
            </ForwardRef(ListItemIcon)>
            <ForwardRef(ListItemText)>
              Total Points Earned: 
              0
            </ForwardRef(ListItemText)>
          </ForwardRef(ListItem)>
        </ForwardRef(List)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
    <ForwardRef(Box)
      sx={
        Object {
          "height": 300,
          "width": "100%",
        }
      }
    >
      <Table
        columns={
          Array [
            Object {
              "field": "name",
              "title": "Name",
            },
            Object {
              "field": "rank",
              "title": "Rank",
              "type": "numeric",
            },
            Object {
              "field": "position",
              "title": "Position",
              "type": "numeric",
            },
            Object {
              "field": "points_earned",
              "title": "Points Earned",
              "type": "numeric",
            },
            Object {
              "field": "percentage_correct",
              "title": "Percentage Correct",
              "type": "numeric",
            },
          ]
        }
        data={Array []}
        icons={
          Object {
            "Add": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Check": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Clear": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Delete": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "DetailPanel": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Edit": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Export": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Filter": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "FirstPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "LastPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "NextPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "PreviousPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "ResetSearch": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Search": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "SortArrow": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "ThirdStateCheck": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "ViewColumn": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        options={
          Object {
            "filtering": true,
          }
        }
        title="Report"
      />
    </ForwardRef(Box)>
  </ForwardRef(Stack)>
</ForwardRef(Box)>
`;
